database/mysql_anonymiser.go

```diff
@@ -21,50 +24,94 @@ func NewMySQLAnonymiser(conn *sql.DB) *MySQLAnonymiser {
+				cell, err := a.anonymiseCell(column, replacement)
+				if err != nil {
+					return err
+				}
+				cells = append(cells, cell)
+			} else {
+				log.Printf("Not marked for anonymising. Skipping cell in column: %v", column)
+				// TODO: Make this approach more efficient.
+				// Currently not very efficient because in each loop,
+				// we create len(columns) number of interfaces so that rows.Scan() will not complain
+				// about not having enough params. in reality, we only need
+				// len(columns) - numberOfAnonymisedFields interfaces
+				nFields := a.shouldNotAnonymise(columns)
+				if err := rows.Scan(nFields...); err != nil {
+					return err
+				}
+				seed := reflect.ValueOf(nFields[idx]).Elem()
+				cell, _ := a.keepsSeedValueUnchanged(column, seed, reflect.TypeOf(seed).Kind())
+				cells = append(cells, cell)
+			}
 		}
 
-		if err := rows.Scan(row...); err != nil {
+		if err := rows.Scan(fields...); err != nil {
 			return err
 		}
 
-		var cells []*Cell
-		for idx, column := range columns {
-			replacement := a.shouldAnonymise(table, column)
-			scanner := row[idx].(*utils.TypeScanner)
-			cells = append(cells, a.anonymiseCell(column, replacement, scanner))
-		}
-
 		rowChan <- cells
 	}
 
 	endChan <- true
 	return nil
 }
 
+// - If the replacement is a literal, then we simply place it in the Cell. No faking here.
+// - If the replacement is not a literal, and it is an actual replacement then call the faker function
+// that it maps on to.
+// - We do not anonymise fields in columns that are not listed in config.
+func (a *MySQLAnonymiser) anonymiseCell(column, replacement string) (*Cell, error) {
+	literal := strings.TrimPrefix(replacement, literalPrefix)
+
+	if a.TypeScanner.Scan(literal); len(literal) != len(replacement) {
+		return &Cell{Column: column, Type: a.Detected, Value: literal}, nil
+	}
+
+	if replacement != "" && len(literal) == len(replacement) {
+		value := (utils.Functions[replacement]).Call([]reflect.Value{})[0]
+		return &Cell{Column: column, Type: a.Detected, Value: value}, nil
+	}
+	return nil, fmt.Errorf("couldn't anonymise cell with column: %v", column)
+}
+
 func (a *MySQLAnonymiser) shouldAnonymise(table, column string) string {
 	return viper.GetString(fmt.Sprintf("anonymise.%s.%s", table, column))
 }
 
-func (a *MySQLAnonymiser) anonymiseCell(column, replacement string, scanner *utils.TypeScanner) *Cell {
-	if replacement != "" && scanner.Detected != "null" {
-		if literal := strings.TrimPrefix(replacement, literalPrefix); len(literal) != len(replacement) {
-			return &Cell{Column: column, Type: scanner.Detected, Value: literal}
-		}
-
-		return &Cell{Column: column, Type: scanner.Detected, Value: utils.Functions[replacement].Call(nil)[0]}
+// Useful for keeping primary key value unchanged. At this point, it is important for us
+// to not include in the config file the primary key column name.
+// TODO: warn user if primary key is listed as anonymised in config file.
```

Comment:
Actually, we'll need a different config item in the file for the primary key. It won't always be called `id`; in fact in the bob DB it never is.

Reply:
Sounds good. Removed the comment. The current approach is probably simpler. We preserve the values of any columns we have not explicitly specified to be anonymised. So a user will not even need extra config for the primary key. :)